feat(data): add GraphQL API optimistic UI docs (#5572)
aws-amplify/docs: src/fragments/lib/graphqlapi/js/optimistic-ui.mdx

The GraphQL API category can be used with [TanStack Query](https://tanstack.com/query/latest/docs/react/overview) to implement optimistic UI, allowing CRUD operations to be rendered immediately on the UI before the request roundtrip has completed. Using the GraphQL API with TanStack additionally makes it easy to render loading and error states, and allows you to rollback changes on the UI when API calls are unsuccessful.

In the following examples we'll create a list view that optimistically renders newly created items, and a detail view that optimistically renders updates and deletes.

<Callout>

For more details on TanStack Query, including requirements, supported browsers, and advanced usage, see the [TanStack Query documentation](https://tanstack.com/query/latest/docs/react/overview). 
For complete guidance on how to implement optimistic updates with TanStack Query, see the [TanStack Query Optimistic UI Documentation](https://tanstack.com/query/latest/docs/react/guides/optimistic-updates).
For more on the Amplify GraphQL API, see the [API documentation](https://docs.amplify.aws/lib/graphqlapi/getting-started/q/platform/js/).

</Callout>

> To get started, run the following command in an existing Amplify project:

Reviewer: ```suggestion
To get started, run the following command in an existing Amplify project with a React frontend:
```

```bash
# Install TanStack Query
npm i @tanstack/react-query

# Select default configuration
amplify add api              
```

> When prompted, use the following schema, which can also be found under `amplify/backend/api/[name of project]/schema.graphql`:

Author: @iartemiev your suggestion is what I previously had written for the API + Storage docs given that the CLI prompts the user to open the schema file for editing, however, @abdallahshaban557 felt strongly that we should include the location here. Marking as resolved, but please feel free to comment if you disagree.

```graphql
type RealEstateProperty @model @auth(rules: [{ allow: public }]) {
  id: ID!
  name: String!
  address: String
}
```

Save the schema and run `amplify push` to deploy the changes. For the purposes of this guide, we'll build a Real Estate Property listing application.

Next, at the root of your project, add the required TanStack Query imports, and create a client:

<Callout>

TanStack Query Devtools are not required, but are a useful resource for debugging and understanding how TanStack works under the hood. By default, React Query Devtools are only included in bundles when `process.env.NODE_ENV === 'development'`, meaning that no additional configuration is required to exclude them from a production build. 
For more information on the TanStack Query Devtools, visit the [TanStack Query Devtools docs](https://tanstack.com/query/latest/docs/react/devtools)

> </Callout>

Reviewer: I'd move this above the "complete examples" call out. It's a bit too prominent up here. It's more useful once the code has been copy/pasted.

Author: I put this here so that the reader understands why we're importing the Devtools in the code snippet that follows. I don't feel strongly either way, so will move.

For more detailed information on query keys, see the [TanStack Query documentation](https://tanstack.com/query/v4/docs/react/guides/query-keys).

## Optimistically rendering a list of records

To optimistically render a list of items returned from the GraphQL API, use the TanStack `useQuery` hook, passing in the GraphQL API query as the `queryFn` parameter. The following example creates a query to retrieve all records from the API. We'll use `realEstateProperties` as the query key, which will be the same key we use to optimistically render a newly created item.

<BlockSwitcher>
<Block name="TypeScript">

```ts
const {
  data: realEstateProperties,
  isLoading,
  isSuccess,
  isError: isErrorQuery,
} = useQuery({
  queryKey: ["realEstateProperties"],
  // TODO: improve return type
  queryFn: async (): Promise<

>     (RealEstateProperty | null | undefined)[] | null | undefined

Reviewer 2: If you omit this type completely does the return type of `realEstateProperties` not get auto-inferred? (Same question applies to all `queryFn` return types later in the doc - can we avoid adding all of those?)

Author: Good callout - the TanStack docs recommended typing the return type of the `queryFn`, so I did it and planned to revisit. Auto-inference works better here! Updated :)

      GraphQLQuery<ListRealEstatePropertiesQuery>
    >({
      query: queries.listRealEstateProperties,
    });

    const allRealEstateProperties =
      response?.data?.listRealEstateProperties?.items;

    // TODO: figure out return type
    if (!allRealEstateProperties) return null;

    return allRealEstateProperties;
  },
});
``` 

</Block>
<Block name="JavaScript">

```js
// TODO - add JS example after approval of TS code snippet
```

</Block>

</BlockSwitcher>

## Optimistically rendering a newly created record

To optimistically render a newly created record returned from the GraphQL API, use the TanStack `useMutation` hook, passing in the GraphQL API mutation as the `mutationFn` parameter. We'll use the same query key used by the `useQuery` hook (`realEstateProperties`) as the query key to optimistically render a newly created item.
We'll use the `onMutate` function to update the cache directly, as well as the `onError` function to rollback changes when a request fails.

<BlockSwitcher>
<Block name="TypeScript">

```ts
const createMutation = useMutation({
mutationFn: async (
  realEstatePropertyDetails: CreateRealEstatePropertyInput

> ): Promise<RealEstateProperty | undefined | null> => {

Reviewer 2: Same question about auto-inferring type

Author: See response above :)

  return isFetching ? <div style={styles.globalLoadingIndicator}></div> : null;
}
```

For more details on advanced usage of TanStack Query hooks, see the [TanStack documentation](https://tanstack.com/query/latest/docs/react/guides/mutations).

The following example demonstrates how to use the state returned by TanStack to render a loading indicator while a mutation is in progress, and an error message if the mutation fails. For additional examples, see the [Complete Example](#complete-example) below.

<BlockSwitcher>
<Block name="TypeScript">

```ts
<>
  {updateMutation.isError &&
  updateMutation.error instanceof Error ? (
    <div>An error occurred: {updateMutation.error.message}</div>
  ) : null}

  {updateMutation.isSuccess ? (
    <div>Real Estate Property updated!</div>
  ) : null}

  <button
    onClick={() =>
      updateMutation.mutate({
        id: realEstateProperty.id,
        address: `${Math.floor(
          1000 + Math.random() * 9000
        )} Main St`,
      })
    }
  >
    Update Address
  </button>
</>
``` 

</Block>
<Block name="JavaScript">

```js
// TODO - add JS example after approval of TS code snippet
```

</Block>

</BlockSwitcher>

> ## Complete examples

Reviewer 3: @david-mcafee Should this just be "Complete example" (singular)?

Author: I made this plural due to the fact that there is both a TS and JS example, but singular makes more sense here!

For more detailed information on query keys, see the [TanStack Query documentation](https://tanstack.com/query/v4/docs/react/guides/query-keys).

## Optimistically rendering a list of records

To optimistically render a list of items returned from the GraphQL API, use the TanStack `useQuery` hook, passing in the GraphQL API query as the `queryFn` parameter. The following example creates a query to retrieve all records from the API. We'll use `realEstateProperties` as the query key, which will be the same key we use to optimistically render a newly created item.

<BlockSwitcher>
<Block name="TypeScript">

```ts
const {
  data: realEstateProperties,
  isLoading,
  isSuccess,
  isError: isErrorQuery,
} = useQuery({
  queryKey: ["realEstateProperties"],
  queryFn: async () => {
    const response = await API.graphql<
      GraphQLQuery<ListRealEstatePropertiesQuery>
    >({
      query: queries.listRealEstateProperties,
    });

    const allRealEstateProperties =
      response?.data?.listRealEstateProperties?.items;

>     // TODO: figure out return type

Author: Removing this!

    ]);

  // Optimistically update to the new value
  if (previousRealEstateProperty) {
    queryClient.setQueryData(
      ["realEstateProperties", newRealEstateProperty.id],
      /**
       * `newRealEstateProperty` will at first only include updated values for
       * the record. To avoid only rendering optimistic values for updated
       * fields on the UI, include the previous values for all fields:
       */
      { ...previousRealEstateProperty, ...newRealEstateProperty }
    );
  }

    // Return a context with the previous and new realEstateProperty
    return { previousRealEstateProperty, newRealEstateProperty };
  },
  // If the mutation fails, use the context we returned above
  onError: (err, newRealEstateProperty, context) => {
    console.error("Error updating record:", err, newRealEstateProperty);
    if (context?.previousRealEstateProperty) {
      queryClient.setQueryData<RealEstateProperty>(
        ["realEstateProperties", context.newRealEstateProperty.id],
        context.previousRealEstateProperty
      );
    }
  },
  // Always refetch after error or success:
  onSettled: (newRealEstateProperty) => {
    queryClient.invalidateQueries({

>       // TODO: fix type

Reviewer: I assume this isn't intended either.

Author: Removing

```js
function GlobalLoadingIndicator() {
  const isFetching = useIsFetching();
  return isFetching ? <div style={styles.globalLoadingIndicator}></div> : null;
}
```

For more details on advanced usage of TanStack Query hooks, see the [TanStack documentation](https://tanstack.com/query/latest/docs/react/guides/mutations).

The following example demonstrates how to use the state returned by TanStack to render a loading indicator while a mutation is in progress, and an error message if the mutation fails. For additional examples, see the [Complete Example](#complete-example) below.

<BlockSwitcher>
<Block name="TypeScript">

```ts
<>
  {updateMutation.isError &&
  updateMutation.error instanceof Error ? (
    <div>An error occurred: {updateMutation.error.message}</div>
  ) : null}

  {updateMutation.isSuccess ? (
    <div>Real Estate Property updated!</div>
  ) : null}

  <button
    onClick={() =>
      updateMutation.mutate({
        id: realEstateProperty.id,
        address: `${Math.floor(
          10c

>         )} Main St`,

Reviewer: this can probably be collapse into 1 line for readabilitiy?

Author: This was accidentally updated - line 629 is missing some math, and if we re-add the code back in and combine with 628 + 630, it will be too long.

The GraphQL API category can be used with [TanStack Query](https://tanstack.com/query/latest/docs/react/overview) to implement optimistic UI, allowing CRUD operations to be rendered immediately on the UI before the request roundtrip has completed.

> To get started, run the following in an existing Amplify project:

Author: This is the same phrasing that we use in the API + Storage docs that were recently added, will update there as well.

```bash
# Install TanStack Query
npm i @tanstack/react-query

# Select default configuration
amplify add api              
```

When prompted, use the following schema, which can also be found under `amplify/backend/api/[name of project]/schema.graphql`:

```graphql
type RealEstateProperty @model @auth(rules: [{ allow: public }]) {
  id: ID!
  name: String!
  address: String
}
```

> Save the schema and run `amplify push` to deploy the changes.

Reviewer: ```suggestion
Save the schema and run `amplify push` to deploy the changes. For the purposes of this guide, we'll build a Real Estate Property listing application.
```

> Next, (at the root of your project?), add the required TanStack Query imports, and create a client:

Reviewer: ```suggestion
Next, in your application code, where you plan to render the real estate properties, add the required TanStack Query imports, and create a client:
```

```ts
import {
  useQuery,
  useMutation,
  useQueryClient,
  QueryClient,
  QueryClientProvider,
} from '@tanstack/react-query'

// Create a client
const queryClient = new QueryClient()
```

Next, provide the client to your app:

```ts
<QueryClientProvider client={queryClient}>
  <App />
</QueryClientProvider>
```

Next, within your app, access the TanStack client:

```ts 
// Access the client
const queryClient = useQueryClient()
```

<Callout>

For more details on TanStack Query, including requirements, supported browsers, and advanced usage, see the [TanStack Query documentation](https://tanstack.com/query/latest/docs/react/overview). 

> For more on Amplify GraphQL API operations, see the [API documentation](https://docs.amplify.aws/lib/graphqlapi/authz/q/platform/js/).

Reviewer: is this the correct link? shouldn't it be this https://docs.amplify.aws/lib/graphqlapi/mutate-data/q/platform/js/ ? Also, I think we should use a relative link here instead of absolute.

import { useIsFetching } from "@tanstack/react-query";

// For rendering 
function GlobalLoadingIndicator() {
  const isFetching = useIsFetching();

  return isFetching ? <Loader size="large" /> : null;
}

// For generating random address
const alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

function App() {
  const [currentRealEstatePropertyId, setCurrentRealEstatePropertyId] =
    useState<string | null>(null);

  // Access the client
  const queryClient = useQueryClient();

  // TanStack Query for listing all realEstateProperties
  const {
    data: realEstateProperties,
    isLoading,
    isSuccess,
    isError: isErrorQuery,
  } = useQuery({
    queryKey: ["realEstateProperties"],
    // TODO: improve return type
    queryFn: async (): Promise<
      (RealEstateProperty | null | undefined)[] | null | undefined
    > => {
      const response = await API.graphql<
        GraphQLQuery<ListRealEstatePropertiesQuery>
      >({
        query: queries.listRealEstateProperties,
      });

      const allRealEstateProperties =
        response?.data?.listRealEstateProperties?.items;

>       if (!allRealEstateProperties) return null;

Reviewer: maybe `return []`

Author: I'll add a todo here - I'll sort out types once we settle on the final approach for the code example.

    onError: (err, newRealEstateProperty, context) => {
      queryClient.setQueryData(
        ["realEstateProperties"],
        // TODO: fix type
        // @ts-ignore
        context.previousRealEstateProperties
      );
    },
    // Always refetch after error or success:
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ["realEstateProperties"] });
    },
  });

  function RealEstatePropertyDetailView() {
    const {
      data: realEstateProperty,
      isLoading,
      isSuccess,
      isError: isErrorQuery,
    } = useQuery({
      queryKey: ["realEstateProperties", currentRealEstatePropertyId],
      // TODO: improve return type
      queryFn: async (): Promise<RealEstateProperty | null | undefined> => {
        const response = await API.graphql<
          GraphQLQuery<GetRealEstatePropertyQuery>
        >({
          query: queries.getRealEstateProperty,
          variables: { id: currentRealEstatePropertyId },
        });

        console.log("RealEstatePropertyDetailView queryFn response", response);
        return response.data?.getRealEstateProperty;
      },
    });

    // TanStack update mutation with optimistic updates

>     const updateMutation = useMutation({

Reviewer: something to verify with TanStack's samples but maybe it can benefit from more use-case-descriptive variable name like "updateRealEstateProperty"

    );

    await response?.data?.listRealEstateProperties?.items.forEach(
      async (realEstateProperty) => {
        if (!realEstateProperty?.id) return;

        const realEstatePropertyDetails: DeleteRealEstatePropertyInput = {
          id: realEstateProperty?.id,
        };

        const deletedRealEstateProperty = await API.graphql<
          GraphQLQuery<DeleteRealEstatePropertyMutation>
        >({
          query: mutations.deleteRealEstateProperty,
          variables: { input: realEstatePropertyDetails },
        });

        console.log("RealEstateProperty deleted: ", deletedRealEstateProperty);
      }
    );
    //endregion

    //region verify all deletes were successful:
    const secondResponse = await API.graphql<
      GraphQLQuery<ListRealEstatePropertiesQuery>
    >({
      query: queries.listRealEstateProperties,
    });
    console.log(
      "RealEstateProperties should be empty:",
      secondResponse?.data?.listRealEstateProperties?.items
    );

    if (secondResponse?.data?.listRealEstateProperties?.items?.length === 0) {
      console.log("All deletes successful!");
    }
    //endregion
  }

  return (
    <div className="App">

>       <header className="App-header">

Reviewer: Is this meant to be `main`?

Author: At this stage I'm only looking for feedback on the TanStack + API logic - I'll finalize styling etc once that's complete, and this is only a placeholder.

        {isErrorQuery && <div>{"Problem loading Real Estate Properties"}</div>}
        {isLoading && <div>{"Loading Real Estate Properties..."}</div>}
        <ul>
          {/* TODO: make this work with isSuccess */}

>           {isSuccess &&

Reviewer: Nit: maybe do an early return for the entire App component if there are errors are isn't succesful.

Author: This is not for error handling, but rather for the TS example: https://tanstack.com/query/v4/docs/react/typescript#type-narrowing

            realEstateProperties?.map((realEstateProperty) => {
              if (!realEstateProperty) return null;
              return (
                <li key={realEstateProperty.id}>
                  {realEstateProperty.name}
                  <button
                    onClick={() =>
                      setCurrentRealEstatePropertyId(realEstateProperty.id)
                    }
                  >
                    Detail View
                  </button>
                </li>
              );
            })}
        </ul>
        <button
          onClick={() => {
            createMutation.mutate({
              name: `New Home ${Date.now()}`,
              address: `${
                alphabet[Math.floor(Math.random() * alphabet.length)]

>               } St`,

Reviewer: Nit: One way to avoid this `alphabet` variable or is to use `window.prompt("...")`. THen customers can provide their own values without us building an entire form interface

Author: I'm trying to keep the UI interface as minimal as possible to highlight the API + TanStack integration, and not confuse readers with anything that's unnecessary. I'll perhaps replace this by updating the number in the address instead, but let me know if you disagree. Also, isolating this to a button click (i.e. excluding user input) allows us to demonstrate optimistic UI for rapid consecutive updates, which I think is useful.
